keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2262 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Feature not used
+
+In a cluster where the feature is not used (no resource driver installed, no
+pods using dynamic resource allocation) the impact is minimal, both for
+performance and security. The scheduler plugin and resource controller will
+return quickly without doing any work for pods.
+
+#### Compromised node
+
+Kubelet is intentionally limited to read-only access for all new API types
+to prevent that a
+compromised kubelet interferes with scheduling of pending pods, for example
+by updating information normally published by the resource driver controller.
+Faking such information could be used for a denial-of-service
+attack against pods using those ResourceClaims, for example by overwriting
+their allocation result with a node selector that matches no node. A
+denial-of-service attack against the cluster and other pods is harder, but
+still possible. For example, frequently updating ResourceClaims could cause new
+scheduling attempts for pending pods. Filling up the `ReservedFor` field
+could exhaust etcd storage.
+
+Another potential attack goal is to get pods with sensitive workloads to run on
+a compromised node. For pods that don't use special resources nothing changes
+in that regard. Such an attack is possible for pods with extended resources
+because kubelet is in control of which capacity it reports for those: it could
+publish much higher values than the device plugin reported and thus attract
+pods to the node that normally would run elsewhere. With dynamic resource
+allocation, such an attack is still possible, but the attack code would have to
+be different for each resource driver because all of them will use their own,
+custom approach for reporting resource availability.
+
+The security of those custom approaches is the responsibility of the resource
+driver vendor. Solutions like Akri which establish their own control plane and
+then communicate with Kubernetes through the device plugin API already need to
+address this.
+
+#### User permissions and quotas
+
+Similar to generic ephemeral inline volumes, the [ephemeral resource use
+case](#ephemeral-vs-persistent-resourceclaims-lifecycle) gets covered by
+creating ResourceClaims on behalf of the user automatically through
+kube-controller-manager. The implication is that RBAC rules that are meant to
+prevent creating ResourceClaims for certain users can be circumvented, at least
+for ephemeral resources. Administrators need to be aware of this caveat when
+designing user restrictions.
+
+A quota system that limits how many resources a user is allowed to allocate
```

Comment:
> https://kubernetes.io/docs/concepts/policy/resource-quotas/#storage-resource-quota

The storage quota system is actually a good example of why doing the quota in Kubernetes for aspects owned by add-ons is frail: in a SIG storage issue, we recently discussed that when a user requests a volume of "1MiB" the actual volume might get rounded up to the minimum volume size supported by the storage backend. Kubernetes then only considers the requested size instead of the actual size of the volume when comparing against the quota.

> This new mechanism should at least support <class>.resourceclass.<somthing?>.k8s.io/resourceclaims ?

We could count the number of resource claims as you suggested. But this looks like a good future enhancement, either of this KEP or even a separate one. Not all storage features where added all at once either.

Reply:
I don't think we can progress to Beta without quota being a solvable problem.  I think alpha should spec "count the number of resource claims of a given class" and beta criteria should be "agreement that quota management is sufficient".

Reply:
I added both.